server/tso/global_allocator.go

```diff
@@ -84,8 +105,173 @@ func (gta *GlobalTSOAllocator) SetTSO(tso uint64) error {
+	}
+	// Send maxTS to all Local TSO Allocator leaders to prewrite
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	maxTSO := &pdpb.Timestamp{}
+	// Collect the MaxTS with all Local TSO Allocator leaders first
+	if err = gta.syncMaxTS(ctx, dcLocationMap, maxTSO); err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	maxTSO.Logical += int64(count)
+	// Sync the MaxTS with all Local TSO Allocator leaders then
+	if err := gta.syncMaxTS(ctx, dcLocationMap, maxTSO); err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	var currentGlobalTSO pdpb.Timestamp
+	if currentGlobalTSO, err = gta.getCurrentTSO(); err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	if tsoutil.CompareTimestamp(&currentGlobalTSO, maxTSO) < 0 {
+		// Update the global TSO in memory
+		if err := gta.SetTSO(tsoutil.GenerateTS(maxTSO)); err != nil {
+			return pdpb.Timestamp{}, err
+		}
+	}
+	return *maxTSO, nil
+}
+
+const (
+	dialTimeout = 3 * time.Second
+	rpcTimeout  = 3 * time.Second
+)
+
+func (gta *GlobalTSOAllocator) syncMaxTS(ctx context.Context, dcLocationMap map[string][]uint64, maxTSO *pdpb.Timestamp) error {
+	maxRetryCount := 1
+	for i := 0; i < maxRetryCount; i++ {
+		// Collect all allocator leaders' client URLs
+		allocatorLeaders, err := gta.allocatorManager.GetLocalAllocatorLeaders()
+		if err != nil {
+			return err
+		}
+		leaderURLs := make([]string, 0, len(allocatorLeaders))
+		for _, allocator := range allocatorLeaders {
+			for _, leaderURL := range allocator.GetMember().GetClientUrls() {
+				if slice.NoneOf(leaderURLs, func(i int) bool { return leaderURLs[i] == leaderURL }) {
+					leaderURLs = append(leaderURLs, leaderURL)
+				}
+			}
+		}
+		// Prepare to make RPC requests concurrently
+		respCh := make(chan *pdpb.SyncMaxTSResponse, len(leaderURLs))
+		errCh := make(chan error, len(leaderURLs))
+		var errList []error
+		wg := sync.WaitGroup{}
+		for _, leaderURL := range leaderURLs {
+			leaderConn, err := gta.getOrCreateGRPCConn(ctx, leaderURL)
+			if err != nil {
+				return err
+			}
+			wg.Add(1)
+			go func(ctx context.Context, conn *grpc.ClientConn, respCh chan<- *pdpb.SyncMaxTSResponse, errCh chan<- error) {
+				request := &pdpb.SyncMaxTSRequest{
+					Header: &pdpb.RequestHeader{
+						SenderId: gta.allocatorManager.member.ID(),
+					},
+				}
+				if maxTSO.GetPhysical() != 0 {
+					request.MaxTs = maxTSO
+				}
+				syncCtx, cancel := context.WithTimeout(ctx, rpcTimeout)
+				resp, err := pdpb.NewPDClient(conn).SyncMaxTS(syncCtx, request)
+				cancel()
+				if err != nil {
+					errCh <- err
+				}
+				respCh <- resp
+				wg.Done()
+			}(ctx, leaderConn, respCh, errCh)
+		}
+		wg.Wait()
+		close(respCh)
+		close(errCh)
+		// If any error occurs, the synchronization process will fail
+		if err := <-errCh; err != nil {
+			errList = append(errList, err)
+		}
+		if len(errList) > 0 {
+			return errs.ErrSyncMaxTS.FastGenWithCause(errList)
+		}
+		var syncedDCs []string
+		for resp := range respCh {
+			if resp == nil {
```

Comment:
Does it always get the nil responds because you close it in line 196?

Reply:
No, it doesn't. I close the channel when all requesting goroutines are done, so if the requests are all successful, the results will keep staying in the buffered channel until I use `for range` to consume them.